src/rt/sched/threadsync.h

```diff
@@ -0,0 +1,101 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+#include "test/systematic.h"
+
+#include <condition_variable>
+#include <mutex>
+
+/**
+ * This file contains the synchronisation implementation for suspending
+ * and resuming threads.  It is design to be a thin wrapper that could
+ * be replaced with alternatives for other platforms.
+ *
+ * This is the standard implementation based on the C++ standard
+ * libraries condition variables.
+ */
+namespace verona::rt
+{
+  template<class T>
+  class ThreadSync
+  {
+    std::mutex m;
+    std::condition_variable cv;
+
+  public:
+    class ThreadSyncHandle
+    {
+      ThreadSync& sync;
+      std::unique_lock<std::mutex> lock;
+      bool wake_on_exit = false;
+
+    public:
+      /**
+       * Wake up all threads in the thread pool
+       *
+       * Will only occur once the handle is dropped.
+       */
+      void unpause_all()
+      {
+        wake_on_exit = true;
+      }
+
+      /**
+       * Pause this thread
+       *
+       * If this is the last thread, then something external must call
+       * unpause_all to restart the paused threads.
+       */
+      void pause()
+      {
+        Systematic::cout() << "Pausing" << Systematic::endl;

```

Comment:
The purpose of this debug message is unclear to me. My understanding is that when systematic testing is enabled, this class is not activated.

Reply:
The `flight_recorder` also logs these messages.  This should really be renamed.  The `flight_recorder` is for debugging crashes that can't be reproduced in systematic testing.